### ISSUE TITLE
Split HTML export in multiple files smaller HTML files

### ISSUE THREAD

**Issue by Skaronator** [bepaald/signalbackup-tools#93]
Currently, a conversation is exported in to a single file. Browsers struggle to load this file when they get really large, especially when there are many images included.

#### Solution

Split the HTML file of a conversation in multiple HTML files. (Maybe 1000 messages per file?) At the end of the first file you have a button "Next messages" which basically loads page2.html. This page contains a "Previous messages" Button on top that loads page1.html and a "Next messages" button at the bottom that loads page3.html.

This is basically the same way how Telegram Desktop handles HTML exports. 

![image](https://user-images.githubusercontent.com/2886913/218104423-eb6eeb00-bc0c-4566-ae0e-4d68f39633ed.png)

Bottom of message.html:
![image](https://user-images.githubusercontent.com/2886913/218104520-805e6fc6-100f-4e84-874e-62ea136cb3b1.png)

Top of message1.html:
![image](https://user-images.githubusercontent.com/2886913/218104586-19c03a04-5e3a-432c-9663-f940dbae6583.png)

Bottom of message1.html:
![image](https://user-images.githubusercontent.com/2886913/218104520-805e6fc6-100f-4e84-874e-62ea136cb3b1.png)

**Comment by bepaald** [comment]
Yes, this was actually always the idea, but I must admit I forgot about it while getting it all working. This will happen, hopefully sometime next week if nothing else comes up.

Thanks for reminding me!

**Comment by Skaronator** [comment]
Awesome, thanks for your work. I've sent you a couple of bucks on PayPal :)

**Comment by bepaald** [comment]
Hey, I just saw that, thank you so much, I really appreciate that!

I also pushed a first version of the splitting code earlier today, it's not (thoroughly) tested yet, so I'm not advertising it (nothing in the readme yet) but you could try it if you want. Just add `--split` to the command line, optionally with a number after it (default to 1000). The number of messages is just a hint by the way, if you split at 1000 messages but a conversation has only 1002, the messages will be divided evenly among the two output files.

I'll probably get around to testing it tomorrow, but if you want to try before that and have some feedback that would be great.

Thanks again!

**Comment by Skaronator** [comment]
Awesome! Just gave it a quick try and already looks really usable! I noticed only one small issue. The avatar picture isn't shown on page 2 and onwards.

Username.html contains the avatar, e.g.:
```html
<label for="zoomCheck-avatar">
<img class="avatar avatar-32 header-avatar msg-sender-32" src="media/Avatar_32.bin" alt="thread avatar">
</label>
```

And Username_1.html and all other only contains the letters:

```html
<div class="avatar header-avatar msg-sender-32">
        N
</div>
```

**Comment by Skaronator** [comment]
Additionally, maybe add navigation on top of each page as well :)

**Comment by bepaald** [comment]
>Awesome! Just gave it a quick try and already looks really usable! I noticed only one small issue. The avatar picture isn't shown on page 2 and onwards.

Yeah, that was stupid, I already suspected this was happening, but I had already left the house, so nothing I could do now. I'll definitely fix it tomorrow. If my thinking is correct, this problem may go away when `append` is added to the command line.

Website design is not my best skill, so this was just a quick idea. Controls at the top would be an improvement indeed. I'm also considering putting the controls at a fixed position at the botton (like the 'back to index' button in the upper left corner) so they're always on view. But I also don't want the page to look too cluttered or messy. Any thoughts on that?

**Comment by Skaronator** [comment]
> Any thoughts on that?

Oh, that would be a good idea. The back and forth button would be always visible left/right side of the chat. Similar to this page: https://pr0gramm.com/top/5544062

I'm not certain how to handle the "first" and "last" page buttons. Maybe just move them outside the chatbox as well but only have them on top/bottom of the page.

Currently:
![image](https://user-images.githubusercontent.com/2886913/218331026-5c4fd36a-4261-4141-95e3-1d30d7f280a9.png)

Possible:
![image](https://user-images.githubusercontent.com/2886913/218331053-41403774-ddc8-4d78-8b62-9cc270517a5b.png)
(But only on top/bottom of the page for "first"/"last" and  "prev"/"next" should be always in view)

**Comment by bepaald** [comment]
So, the avatar problem is fixed I think.

For the controls I went with a slightly different solution: a menu in place of the previous 'back to index' button. It is always in view and (since it expands on hover) doesn't require any extra clicks while still keeping the page fairly clean (it's still only one element). Also, it is extendible if I think of more page-controls in the future (maybe a switch dark/light theme button?*) without ever growing.

I'm still not sure though, I might decide at a later point to go back to the previous design, your suggestion or even something completely new. Let me know if you have a strong opinion about it. And if the export now works well, with splitting and avatars and everything.

Thanks!

\* though I think that would require javascript which I'd rather avoid.

**Comment by Skaronator** [comment]
Avatar problem is fixed, yes!

I don't really like the new UI. It is a bit clumsy when you have to skip 3 or 4 pages at once.

I did a quick prototype that would be way more UX friendly.

![image](https://user-images.githubusercontent.com/2886913/218516235-d0228290-0766-4c99-afb0-e5689310ff21.png)

Here is the source code:
[Prototype.zip](https://github.com/bepaald/signalbackup-tools/files/10724221/Prototype.zip)

The buttons are sticky and always in view. I didn't cleanup the css from the old menu.

**Comment by bepaald** [comment]
Yes, I think the menu looked more elegant but I guess when you actually want to _use_ the navigation, it's nice if you easily can :-)

So, I went with your suggestion... mostly. I did move the buttons up to be on line with the 'index' button (the place where they would stick after any amount of scrolling anyway). Do you think it's ok like this?

**Comment by bepaald** [comment]
Maybe having them stuck at the bottom is even nicer?

![screenshot4](https://user-images.githubusercontent.com/38437099/218714423-0b7fc480-c66c-4a72-bd74-2ec76600ee31.png)

**Comment by Skaronator** [comment]
Yes! Bottom looks way better :)

**Comment by bepaald** [comment]
Done!

Thanks for all your help with this one.

**Comment by Skaronator** [comment]
Just gave it a try and looks wonderful! Thanks :)

I noticed a small issue with really long links:

![image](https://user-images.githubusercontent.com/2886913/218742216-16ae052f-52ac-496e-bbe4-8e555e3e142f.png)

Quick fix that works for me:

```diff
.msg {
        max-width: 50%;
        border-radius: .6em;
        margin: 7px 0;
        padding: 10px;
        position: relative;
+       overflow-wrap: break-word;
+       word-wrap: break-word;
+       word-break: break-word;
}
```

Source: https://css-tricks.com/snippets/css/prevent-long-urls-from-breaking-out-of-container/

**Comment by Skaronator** [comment]
Just gave it another try and found some new issues with really long links :( I think we have to use `overflow-wrap: anywhere;` to avoid overflowing at all cost.

![image](https://user-images.githubusercontent.com/2886913/218776070-01ee80d3-84a5-492d-b30e-910726bd24e1.png)

with `overflow-wrap: anywhere;`

![image](https://user-images.githubusercontent.com/2886913/218776231-4f2349e1-9dfa-4def-9639-a54c288ccc67.png)


Other than that, it looks perfect! :heart:

**Comment by bepaald** [comment]
I tried that one too, but things seemed to jump around when I did that. Turns out: they jumped to the correct place! Haha. the very very long line + break-word was causing things to move off-center.

Should be done now. I'm starting to hope that's the last of it, CSS is very satisfying when it does what you want (aka when you know what you're doing), but it can be frustrating when you don't understand it all (like me).

But, seriously, if you find more I'd of course love to hear it.
